core/test/Cardano/Address/ScriptSpec.hs

```diff
@@ -296,28 +331,158 @@ spec = do
                     ]
             validateScript script `shouldBe` Right ()
 
-    describe "can perform roundtrip JSON serialization & deserialization" $
-        it "fromJSON . toJSON === pure" $ property prop_jsonRoundtrip
+    describe "validateScriptTemplate - errors" $ do
+        let accXpub0 =
+                "7eebe6dfa9a1530248400eb6a1adaca166ab1d723e9618d989d22a9219a364\
+                \cb4c745e128fdc98a5039893f704cf67f58c59cea97241a5c7ec7b4606253e5523"
+        let accXpub1 =
+                "417236c94b3ad73557a4df690527f77bebd203de7a208fb3be9c5efa675aaa\
+                \967ca13a50a2f2e95364d0b7fdc75a82e8cc97b499ecd6b9ba12529dd63a2ca7d5"
+        let accXpub2 =
+                "ebf69a16263b741240d3a3d67b44be3a70516adc1a7422b214d0e379314692\
+                \9eb053c9d5500fdcc4088b6a2c3b20b145d84ca77d5ad59343ddf4ba6c9b482d7c"
+        let accXpub3 =
+                "30a71e7919e9c409811efe8d818b831096ac44678397e8911c921a19f2e9b7\
+                \f45b45a93ec2432ed0d314e356a69409c21823f152ae898a97b9b6f72ecd9c2400"
+        let cosigners' = Map.fromList $
+                zipWith (\ix accXpub -> (Cosigner ix, encodeXPubFromTxtUnsafe accXpub))
+                [0, 1, 2, 3] [accXpub0, accXpub1, accXpub2, accXpub3]
+        let cosignersWrong = Map.fromList $
+                zipWith (\ix accXpub -> (Cosigner ix, encodeXPubFromTxtUnsafe accXpub))
+                [0, 1, 2, 3] [accXpub0, accXpub1, accXpub2, accXpub0]
+        let cosigner0 = RequireSignatureOf (Cosigner 0)
+        let cosigner1 = RequireSignatureOf (Cosigner 1)
+        let cosigner2 = RequireSignatureOf (Cosigner 2)
+        let cosigner3 = RequireSignatureOf (Cosigner 3)
+
+        it "no cosigners in script template" $ do
+            let scriptTemplate = ScriptTemplate Map.empty (RequireAllOf [])
+            validateScriptTemplate scriptTemplate `shouldBe` (Left NoCosigner)
+
+        it "illegal cosigner in script template" $ do
+            let scriptTemplate = ScriptTemplate cosigners' (RequireSignatureOf (Cosigner 4))
+            validateScriptTemplate scriptTemplate `shouldBe` (Left UnknownCosigner)
```

Comment:
I'd like to see the other way around too, an extra cosigner in the map that is not mentioned in the script ;)

Reply:
done